src/heap_utils.cc

```diff
@@ -231,19 +234,178 @@ class BufferOutputStream : public v8::OutputStream {
+  }
+
+  int GetChunkSize() override {
+    return 65536;  // big chunks == faster
+  }
+
+  void EndOfStream() override {
+    ended_ = true;
+    EmitRead(UV_EOF);
+  }
+
+  WriteResult WriteAsciiChunk(char* data, int size) override {
+    uv_buf_t buf = EmitAlloc(size);
+    memcpy(buf.base, data, size);
+    EmitRead(size, buf);
+    return kContinue;
+  }
+
+  int ReadStart() override {
+    snapshot_->Serialize(this, HeapSnapshot::kJSON);
+    return 0;
+  }
+
+  int ReadStop() override {
+    return 0;
+  }
+
+  int DoShutdown(ShutdownWrap* req_wrap) override {
+    UNREACHABLE();
+    return 0;
+  }
+
+  int DoWrite(WriteWrap* w,
+              uv_buf_t* bufs,
+              size_t count,
+              uv_stream_t* send_handle) override {
+    UNREACHABLE();
+    return 0;
+  }
+
+  bool IsAlive() override { return !ended_; }
+  bool IsClosing() override { return ended_; }
+  AsyncWrap* GetAsyncWrap() override { return this; }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {}
+
+  SET_MEMORY_INFO_NAME(StreamBaseOutputStream)
+  SET_SELF_SIZE(StreamBaseOutputStream)
+
+ private:
+  const HeapSnapshot* snapshot_;
+  bool ended_;
+};
+
+StreamBaseOutputStream* NewStreamBaseOutputStream(
+    Environment* env,
+    const HeapSnapshot* const snapshot) {
+  Local<Object> obj;
+  if (!env->streambaseoutputstream_constructor_template()
+           ->NewInstance(env->context())
+           .ToLocal(&obj)) {
+    return nullptr;
+  }
+  return new StreamBaseOutputStream(env, snapshot, obj);
+}
+
+inline void TakeSnapshot(Isolate* isolate, v8::OutputStream* out) {
+  const HeapSnapshot* const snapshot =
+      isolate->GetHeapProfiler()->TakeHeapSnapshot();
+  snapshot->Serialize(out, HeapSnapshot::kJSON);
+  const_cast<HeapSnapshot*>(snapshot)->Delete();
+}
+
+inline bool WriteSnapshot(Isolate* isolate, const char* filename) {
+  FILE* fp = fopen(filename, "w");
+  if (fp == nullptr)
+    return false;
+  FileOutputStream stream(fp);
+  TakeSnapshot(isolate, &stream);
+  fclose(fp);
+  return true;
+}
+
+}  // namespace
+
 void CreateHeapDump(const FunctionCallbackInfo<Value>& args) {
   Isolate* isolate = args.GetIsolate();
-  const HeapSnapshot* snapshot = isolate->GetHeapProfiler()->TakeHeapSnapshot();
   BufferOutputStream out;
-  snapshot->Serialize(&out, HeapSnapshot::kJSON);
-  const_cast<HeapSnapshot*>(snapshot)->Delete();
+  TakeSnapshot(isolate, &out);
   Local<Value> ret;
   if (JSON::Parse(isolate->GetCurrentContext(),
                   out.ToString(isolate)).ToLocal(&ret)) {
     args.GetReturnValue().Set(ret);
   }
 }
 
+void CreateHeapDumpStream(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  HandleScope scope(env->isolate());
+  const HeapSnapshot* const snapshot =
+      env->isolate()->GetHeapProfiler()->TakeHeapSnapshot();
+  StreamBaseOutputStream* out = NewStreamBaseOutputStream(env, snapshot);
+  args.GetReturnValue().Set(out->object());
```

Comment:
Needs a nullptr check because `out` can be nullptr.

You could inline NewStreamBaseOutputStream() into this function, it's the only caller.